File: quickmaths/distributions.py
"""Dependency-free, bounded probability distribution helpers.

These functions are used only by trusted native lesson generation.  The
continued fractions and series below are deliberately bounded so malformed or
pathological authored values cannot turn evaluation into an unbounded loop.
"""

from __future__ import annotations

import math
from numbers import Real


_MAX_ITERATIONS = 300
_RELATIVE_TOLERANCE = 1e-14
_FPMIN = 1e-300
_SQRT_2 = math.sqrt(2.0)


def _finite_real(value: object, name: str) -> float:
    if isinstance(value, bool) or not isinstance(value, Real):
        raise ValueError(f"{name} must be a finite real number")
    result = float(value)
    if not math.isfinite(result):
        raise ValueError(f"{name} must be a finite real number")
    return result


def _positive(value: object, name: str) -> float:
    result = _finite_real(value, name)
    if result <= 0.0:
        raise ValueError(f"{name} must be greater than zero")
    return result


def _probability(value: float) -> float:
    # Round only tiny floating point excursions at the closed probability
    # boundaries.  The algorithms otherwise return a value in [0, 1].
    if not math.isfinite(value):
        raise ValueError("probability computation did not converge")
    if value <= 0.0:
        return 0.0
    if value >= 1.0:
        return 1.0
    return value


def normal_cdf(z: object) -> float:
    """Return the lower-tail standard normal probability at finite ``z``."""
    value = _finite_real(z, "z")
    return _probability(0.5 * math.erfc(-value / _SQRT_2))


def inverse_normal_cdf(p: object) -> float:
    """Return the standard normal quantile for a probability strictly in (0, 1)."""
    probability = _finite_real(p, "p")
    if not 0.0 < probability < 1.0:
        raise ValueError("p must be strictly between zero and one")

    # Peter J. Acklam's rational approximation, followed by one Halley step.
    # The approximation is evaluated in the tail that is numerically safest.
    a = (-3.969683028665376e1, 2.209460984245205e2,
         -2.759285104469687e2, 1.383577518672690e2,
         -3.066479806614716e1, 2.506628277459239)
    b = (-5.447609879822406e1, 1.615858368580409e2,
         -1.556989798598866e2, 6.680131188771972e1,
         -1.328068155288572e1)
    c = (-7.784894002430293e-3, -3.223964580411365e-1,
         -2.400758277161838, -2.549732539343734,
         4.374664141464968, 2.938163982698783)
    d = (7.784695709041462e-3, 3.224671290700398e-1,
         2.445134137142996, 3.754408661907416)
    low = 0.02425
    high = 1.0 - low
    def horner(coefficients: tuple[float, ...], argument: float) -> float:
        result = coefficients[0]
        for coefficient in coefficients[1:]:
            result = result * argument + coefficient
        return result

    if probability < low:
        q = math.sqrt(-2.0 * math.log(probability))
        value = horner(c, q) / horner((*d, 1.0), q)
    elif probability <= high:
        q = probability - 0.5
        r = q * q
        value = horner(a, r) * q / horner((*b, 1.0), r)
    else:
        q = math.sqrt(-2.0 * math.log1p(-probability))
        value = -horner(c, q) / horner((*d, 1.0), q)

    # Halley's correction materially improves parity with reference libraries
    # around the centre while retaining the accurate tail approximation.
    density = math.exp(-0.5 * value * value) / math.sqrt(2.0 * math.pi)
    if density > 0.0 and math.isfinite(density):
        if value >= 0.0:
            # Compute the upper tail directly; subtracting a CDF from one
            # loses the useful digits in the upper half of the distribution.
            error = 0.5 * math.erfc(value / _SQRT_2) - (1.0 - probability)
            value += error / density / (1.0 + 0.5 * value * error / density)
        else:
            error = normal_cdf(value) - probability
            value -= error / density / (1.0 + 0.5 * value * error / density)
    return value


def _log_prefactor(a: float, x: float) -> float:
    try:
        log_gamma = math.lgamma(a)
    except OverflowError:
        raise ValueError("incomplete gamma parameters exceed stable range")
    exponent = -x + a * math.log(x) - log_gamma
    if math.isnan(exponent) or exponent == math.inf:
        raise ValueError("incomplete gamma parameters exceed stable range")
    return exponent


def _exp_log_prefactor(a: float, x: float) -> float:
    exponent = _log_prefactor(a, x)
    if exponent <= -745.0:
        return 0.0
    if exponent >= 709.0:
        return math.inf
    return math.exp(exponent)


def _regularized_gamma(a: float, x: float) -> tuple[float, float]:
    """Return (P(a,x), Q(a,x)) with bounded series/continued fraction."""
    if x == 0.0:
        return 0.0, 1.0
    if x < a + 1.0:
        # Track a-scaled terms so log(P) can use lgamma(a+1) directly.  This
        # avoids subtracting two ~log(a) values and preserves Q for tiny df.
        scaled_total = 1.0
        scaled_term = 1.0
        for index in range(1, _MAX_ITERATIONS + 1):
            scaled_term *= x / (a + index)
            scaled_total += scaled_term
            if abs(scaled_term) <= abs(scaled_total) * _RELATIVE_TOLERANCE:
                try:
                    log_lower = -x + a * math.log(x) - math.lgamma(a + 1.0) + math.log(scaled_total)
                except OverflowError as error:
                    raise ValueError("incomplete gamma parameters exceed stable range") from error
                if math.isnan(log_lower) or log_lower == math.inf:
                    raise ValueError("incomplete gamma parameters exceed stable range")
                if log_lower >= 0.0:
                    return 1.0, 0.0
                lower = 0.0 if log_lower <= -745.0 else math.exp(log_lower)
                # -expm1(log(P)) retains the upper-tail digits when P is
                # close to one (especially for very small fractional df).
                upper = -math.expm1(log_lower)
                return _probability(lower), _probability(upper)
        raise ValueError("incomplete gamma series did not converge")

    b = x + 1.0 - a
    d = 1.0 / b if b != 0.0 else 1.0 / _FPMIN
    c = 1.0 / _FPMIN
    h = d
    for index in range(1, _MAX_ITERATIONS + 1):
        coefficient = -index * (index - a)
        b += 2.0
        d = coefficient * d + b
        if abs(d) < _FPMIN:
            d = _FPMIN
        c = b + coefficient / c
        if abs(c) < _FPMIN:
            c = _FPMIN
        d = 1.0 / d
        delta = d * c
        h *= delta
        if abs(delta - 1.0) <= _RELATIVE_TOLERANCE:
            upper = _exp_log_prefactor(a, x) * h
            upper = _probability(upper)
            return _probability(1.0 - upper), upper
    raise ValueError("incomplete gamma continued fraction did not converge")


def chi_square_cdf(x: object, df: object) -> float:
    """Return the lower-tail chi-square probability."""
    value = _finite_real(x, "x")
    if value < 0.0:
        raise ValueError("x must be non-negative")
    degrees = _positive(df, "df")
    return _regularized_gamma(degrees * 0.5, value * 0.5)[0]


def chi_square_sf(x: object, df: object) -> float:
    """Return the chi-square upper tail directly via Q(df/2, x/2)."""
    value = _finite_real(x, "x")
    if value < 0.0:
        raise ValueError("x must be non-negative")
    degrees = _positive(df, "df")
    return _regularized_gamma(degrees * 0.5, value * 0.5)[1]


def _beta_continued_fraction(a: float, b: float, x: float) -> float:
    qab = a + b
    qap = a + 1.0
    qam = a - 1.0
    c = 1.0
    d = 1.0 - qab * x / qap
    if abs(d) < _FPMIN:
        d = _FPMIN
    d = 1.0 / d
    result = d
    for index in range(1, _MAX_ITERATIONS + 1):
        m2 = 2.0 * index
        numerator = index * (b - index) * x / ((qam + m2) * (a + m2))
        d = 1.0 + numerator * d
        if abs(d) < _FPMIN:
            d = _FPMIN
        c = 1.0 + numerator / c
        if abs(c) < _FPMIN:
            c = _FPMIN
        d = 1.0 / d
        result *= d * c
        numerator = -(a + index) * (qab + index) * x / ((a + m2) * (qap + m2))
        d = 1.0 + numerator * d
        if abs(d) < _FPMIN:
            d = _FPMIN
        c = 1.0 + numerator / c
        if abs(c) < _FPMIN:
            c = _FPMIN
        d = 1.0 / d
        delta = d * c
        result *= delta
        if abs(delta - 1.0) <= _RELATIVE_TOLERANCE:
            return result
    raise ValueError("incomplete beta continued fraction did not converge")


def _regularized_beta(x: float, a: float, b: float) -> float:
    if x <= 0.0:
        return 0.0
    if x >= 1.0:
        return 1.0
    log_term = (math.lgamma(a + b) - math.lgamma(a) - math.lgamma(b)
                + a * math.log(x) + b * math.log1p(-x))
    term = 0.0 if log_term <= -745.0 else math.exp(log_term)
    threshold = (a + 1.0) / (a + b + 2.0)
    if x < threshold:
        return _probability(term * _beta_continued_fraction(a, b, x) / a)
    complement = term * _beta_continued_fraction(b, a, 1.0 - x) / b
    return _probability(1.0 - complement)


def _regularized_beta_log_x(log_x: float, a: float, b: float) -> float:
    """Small-x beta tail when x itself underflows a binary64 value."""
    try:
        gamma_difference = math.lgamma(a + b) - math.lgamma(b)
    except OverflowError as error:
        raise ValueError("incomplete beta parameters exceed stable range") from error
    log_term = gamma_difference - math.lgamma(a) + a * log_x
    if log_term <= -745.0:
        return 0.0
    return _probability(math.exp(log_term) / a)


def t_cdf(t: object, df: object) -> float:
    """Return the lower-tail Student t probability for finite t and df > 0."""
    value = _finite_real(t, "t")
    degrees = _positive(df, "df")
    if value == 0.0:
        return 0.5
    # Keep df/(df+t²) in log space: t² may overflow although the beta tail
    # remains representable (notably for tiny fractional df).
    log_degrees = math.log(degrees)
    log_squared = 2.0 * math.log(abs(value))
    largest = max(log_degrees, log_squared)
    log_denominator = largest + math.log1p(math.exp(min(log_degrees, log_squared) - largest))
    log_ratio = log_degrees - log_denominator
    if log_ratio <= -745.0:
        beta_tail = _regularized_beta_log_x(log_ratio, degrees * 0.5, 0.5)
    else:
        beta_tail = _regularized_beta(math.exp(log_ratio), degrees * 0.5, 0.5)
    tail = 0.5 * beta_tail
    return _probability(tail if value < 0.0 else 1.0 - tail)


def f_sf(x: object, df1: object, df2: object) -> float:
    """Return the upper-tail F probability via I(df2/(df2+df1*x))."""
    value = _finite_real(x, "x")
    if value < 0.0:
        raise ValueError("x must be non-negative")
    first = _positive(df1, "df1")
    second = _positive(df2, "df2")
    if value == 0.0:
        return 1.0
    # Form z in log space so df1*x overflow cannot turn a representable
    # upper tail into an artificial zero.
    log_second = math.log(second)
    log_product = math.log(first) + math.log(value)
    largest = max(log_second, log_product)
    log_denominator = largest + math.log1p(math.exp(min(log_second, log_product) - largest))
    log_ratio = log_second - log_denominator
    if log_ratio <= -745.0:
        return _regularized_beta_log_x(log_ratio, second * 0.5, first * 0.5)
    return _regularized_beta(math.exp(log_ratio), second * 0.5, first * 0.5)


__all__ = [
    "normal_cdf", "inverse_normal_cdf", "t_cdf",
    "chi_square_cdf", "chi_square_sf", "f_sf",
]

File: docs/distributions.js
// Dependency-free, bounded distribution kernels for trusted native templates.
// Keep coefficient order and stopping rules aligned with quickmaths/distributions.py.

const MAX_ITERATIONS = 300;
const RELATIVE_TOLERANCE = 1e-14;
const FPMIN = 1e-300;
const SQRT_2 = Math.sqrt(2);

function finiteReal(value, name) {
  if (typeof value !== "number" || !Number.isFinite(value)) throw new Error(`${name} must be a finite real number.`);
  return value;
}

function positive(value, name) {
  const result = finiteReal(value, name);
  if (!(result > 0)) throw new Error(`${name} must be greater than zero.`);
  return result;
}

function probability(value) {
  if (!Number.isFinite(value)) throw new Error("probability computation did not converge.");
  if (value <= 0) return 0;
  if (value >= 1) return 1;
  return value;
}

export function normal_cdf(z) {
  if (arguments.length !== 1) throw new Error("normal_cdf expects one argument.");
  const value = finiteReal(z, "z");
  if (value <= -39) return 0;
  if (value >= 39) return 1;
  const x = Math.abs(value) / SQRT_2;
  const tail = 0.5 * regularizedGamma(0.5, x * x)[1];
  return probability(value < 0 ? tail : 1 - tail);
}

// Accurate complementary error function using the same normal-tail continued
// fraction as the Python implementation.  This avoids the ~1e-7 A&S error in
// the acceptance vectors while remaining dependency-free.
function erfc(value) {
  const x = finiteReal(value, "x");
  if (x < 0) return 2 - erfc(-x);
  return regularizedGamma(0.5, x * x)[1];
}

export function inverse_normal_cdf(p) {
  if (arguments.length !== 1) throw new Error("inverse_normal_cdf expects one argument.");
  const probabilityValue = finiteReal(p, "p");
  if (!(probabilityValue > 0 && probabilityValue < 1)) throw new Error("p must be strictly between zero and one.");
  const a = [-39.69683028665376, 220.9460984245205, -275.9285104469687, 138.357751867269, -30.66479806614716, 2.506628277459239];
  const b = [-54.47609879822406, 161.5858368580409, -155.6989798598866, 66.80131188771972, -13.28068155288572];
  const c = [-0.007784894002430293, -0.3223964580411365, -2.400758277161838, -2.549732539343734, 4.374664141464968, 2.938163982698783];
  const d = [0.007784695709041462, 0.3224671290700398, 2.445134137142996, 3.754408661907416];
  const horner = (coefficients, argument) => coefficients.slice(1).reduce((result, coefficient) => result * argument + coefficient, coefficients[0]);
  const low = 0.02425;
  const high = 1 - low;
  let value;
  if (probabilityValue < low) {
    const q = Math.sqrt(-2 * Math.log(probabilityValue));
    value = horner(c, q) / horner([...d, 1], q);
  } else if (probabilityValue <= high) {
    const q = probabilityValue - 0.5;
    value = horner(a, q * q) * q / horner([...b, 1], q * q);
  } else {
    const q = Math.sqrt(-2 * Math.log1p(-probabilityValue));
    value = -horner(c, q) / horner([...d, 1], q);
  }
  const density = Math.exp(-0.5 * value * value) / Math.sqrt(2 * Math.PI);
  if (density > 0 && Number.isFinite(density)) {
    const error = value >= 0
      ? 0.5 * erfc(value / SQRT_2) - (1 - probabilityValue)
      : normal_cdf(value) - probabilityValue;
    value += value >= 0
      ? error / density / (1 + 0.5 * value * error / density)
      : -error / density / (1 + 0.5 * value * error / density);
  }
  return value;
}

function logGamma(value) {
  const coefficients = [0.9999999999998099, 676.5203681218851, -1259.1392167224028, 771.3234287776531, -176.6150291621406, 12.50734327868691, -0.13857109526572012, 9.984369578019572e-6, 1.5056327351493116e-7];
  if (value < 0.5) return Math.log(Math.PI) - Math.log(Math.sin(Math.PI * value)) - logGamma(1 - value);
  const shifted = value - 1;
  let sum = coefficients[0];
  for (let index = 1; index < coefficients.length; index += 1) sum += coefficients[index] / (shifted + index);
  const t = shifted + 7.5;
  return 0.5 * Math.log(2 * Math.PI) + (shifted + 0.5) * Math.log(t) - t + Math.log(sum);
}

function logPrefactor(a, x) {
  const logGammaValue = logGamma(a);
  if (!Number.isFinite(logGammaValue)) throw new Error("incomplete gamma parameters exceed stable range.");
  const exponent = -x + a * Math.log(x) - logGammaValue;
  if (Number.isNaN(exponent) || exponent === Infinity) throw new Error("incomplete gamma parameters exceed stable range.");
  return exponent;
}

function expLogPrefactor(a, x) {
  const exponent = logPrefactor(a, x);
  if (exponent <= -745) return 0;
  if (exponent >= 709) return Infinity;
  return Math.exp(exponent);
}

function regularizedGamma(a, x) {
  if (x === 0) return [0, 1];
  if (x < a + 1) {
    // Keep a-scaled terms and use lgamma(a+1), preserving Q for tiny df.
    let scaledTotal = 1;
    let scaledTerm = 1;
    for (let index = 1; index <= MAX_ITERATIONS; index += 1) {
      scaledTerm *= x / (a + index);
      scaledTotal += scaledTerm;
      if (Math.abs(scaledTerm) <= Math.abs(scaledTotal) * RELATIVE_TOLERANCE) {
        const logGammaNext = logGamma(a + 1);
        if (!Number.isFinite(logGammaNext)) throw new Error("incomplete gamma parameters exceed stable range.");
        const logLower = -x + a * Math.log(x) - logGammaNext + Math.log(scaledTotal);
        if (Number.isNaN(logLower) || logLower === Infinity) throw new Error("incomplete gamma parameters exceed stable range.");
        if (logLower >= 0) return [1, 0];
        const lower = logLower <= -745 ? 0 : Math.exp(logLower);
        return [probability(lower), probability(-Math.expm1(logLower))];
      }
    }
    throw new Error("incomplete gamma series did not converge.");
  }
  let b = x + 1 - a;
  let d = b === 0 ? 1 / FPMIN : 1 / b;
  let c = 1 / FPMIN;
  let h = d;
  for (let index = 1; index <= MAX_ITERATIONS; index += 1) {
    const coefficient = -index * (index - a);
    b += 2;
    d = coefficient * d + b;
    if (Math.abs(d) < FPMIN) d = FPMIN;
    c = b + coefficient / c;
    if (Math.abs(c) < FPMIN) c = FPMIN;
    d = 1 / d;
    const delta = d * c;
    h *= delta;
    if (Math.abs(delta - 1) <= RELATIVE_TOLERANCE) {
      const upper = probability(expLogPrefactor(a, x) * h);
      return [probability(1 - upper), upper];
    }
  }
  throw new Error("incomplete gamma continued fraction did not converge.");
}

export function chi_square_cdf(x, df) {
  if (arguments.length !== 2) throw new Error("chi_square_cdf expects two arguments.");
  const value = finiteReal(x, "x");
  if (value < 0) throw new Error("x must be non-negative.");
  const degrees = positive(df, "df");
  return regularizedGamma(degrees * 0.5, value * 0.5)[0];
}

export function chi_square_sf(x, df) {
  if (arguments.length !== 2) throw new Error("chi_square_sf expects two arguments.");
  const value = finiteReal(x, "x");
  if (value < 0) throw new Error("x must be non-negative.");
  const degrees = positive(df, "df");
  return regularizedGamma(degrees * 0.5, value * 0.5)[1];
}

function betaContinuedFraction(a, b, x) {
  const qab = a + b;
  const qap = a + 1;
  const qam = a - 1;
  let c = 1;
  let d = 1 - qab * x / qap;
  if (Math.abs(d) < FPMIN) d = FPMIN;
  d = 1 / d;
  let result = d;
  for (let index = 1; index <= MAX_ITERATIONS; index += 1) {
    const m2 = 2 * index;
    let numerator = index * (b - index) * x / ((qam + m2) * (a + m2));
    d = 1 + numerator * d;
    if (Math.abs(d) < FPMIN) d = FPMIN;
    c = 1 + numerator / c;
    if (Math.abs(c) < FPMIN) c = FPMIN;
    d = 1 / d;
    result *= d * c;
    numerator = -(a + index) * (qab + index) * x / ((a + m2) * (qap + m2));
    d = 1 + numerator * d;
    if (Math.abs(d) < FPMIN) d = FPMIN;
    c = 1 + numerator / c;
    if (Math.abs(c) < FPMIN) c = FPMIN;
    d = 1 / d;
    const delta = d * c;
    result *= delta;
    if (Math.abs(delta - 1) <= RELATIVE_TOLERANCE) return result;
  }
  throw new Error("incomplete beta continued fraction did not converge.");
}

function regularizedBeta(x, a, b) {
  if (x <= 0) return 0;
  if (x >= 1) return 1;
  const logTerm = logGamma(a + b) - logGamma(a) - logGamma(b) + a * Math.log(x) + b * Math.log1p(-x);
  const term = logTerm <= -745 ? 0 : Math.exp(logTerm);
  const threshold = (a + 1) / (a + b + 2);
  if (x < threshold) return probability(term * betaContinuedFraction(a, b, x) / a);
  return probability(1 - term * betaContinuedFraction(b, a, 1 - x) / b);
}

function regularizedBetaLogX(logX, a, b) {
  let gammaDifference = logGamma(a + b) - logGamma(b);
  if (!Number.isFinite(gammaDifference)) {
    throw new Error("incomplete beta parameters exceed stable range.");
  }
  const logTerm = gammaDifference - logGamma(a) + a * logX;
  if (logTerm <= -745) return 0;
  return probability(Math.exp(logTerm) / a);
}

export function t_cdf(t, df) {
  if (arguments.length !== 2) throw new Error("t_cdf expects two arguments.");
  const value = finiteReal(t, "t");
  const degrees = positive(df, "df");
  if (value === 0) return 0.5;
  const logDegrees = Math.log(degrees);
  const logSquared = 2 * Math.log(Math.abs(value));
  const largest = Math.max(logDegrees, logSquared);
  const logDenominator = largest + Math.log1p(Math.exp(Math.min(logDegrees, logSquared) - largest));
  const logRatio = logDegrees - logDenominator;
  const betaTail = logRatio <= -745
    ? regularizedBetaLogX(logRatio, degrees * 0.5, 0.5)
    : regularizedBeta(Math.exp(logRatio), degrees * 0.5, 0.5);
  const tail = 0.5 * betaTail;
  return probability(value < 0 ? tail : 1 - tail);
}

export function f_sf(x, df1, df2) {
  if (arguments.length !== 3) throw new Error("f_sf expects three arguments.");
  const value = finiteReal(x, "x");
  if (value < 0) throw new Error("x must be non-negative.");
  const first = positive(df1, "df1");
  const second = positive(df2, "df2");
  if (value === 0) return 1;
  const logSecond = Math.log(second);
  const logProduct = Math.log(first) + Math.log(value);
  const largest = Math.max(logSecond, logProduct);
  const logDenominator = largest + Math.log1p(Math.exp(Math.min(logSecond, logProduct) - largest));
  const logRatio = logSecond - logDenominator;
  if (logRatio <= -745) return regularizedBetaLogX(logRatio, second * 0.5, first * 0.5);
  return regularizedBeta(Math.exp(logRatio), second * 0.5, first * 0.5);
}
